libraries/app/api.cpp

```diff
@@ -304,23 +304,26 @@ namespace graphene { namespace app {
        vector<operation_history_object> result;
        const auto& stats = account(db).statistics(db);
        if( stats.most_recent_op == account_transaction_history_id_type() ) return result;
+       if( start.instance.value < stop.instance.value) return result;
+       if( start.instance.value == stop.instance.value && start != operation_history_id_type()) return result;
        const account_transaction_history_object* node = &stats.most_recent_op(db);
        if( start == operation_history_id_type() )
           start = node->operation_id;
 
-       while(node && node->operation_id.instance.value > stop.instance.value && result.size() < limit)
-       {
-          if( node->operation_id.instance.value <= start.instance.value )
-             result.push_back( node->operation_id(db) );
-          if( node->next == account_transaction_history_id_type() )
-             node = nullptr;
-          else node = &node->next(db);
-       }
-       if( stop.instance.value == 0 && result.size() < limit )
+       const auto& hist_idx = db.get_index_type<account_transaction_history_index>();
+       const auto& by_op_idx = hist_idx.indices().get<by_op>();
+
+       auto index_end = by_op_idx.end();
+
+       auto itr = by_op_idx.upper_bound(boost::make_tuple(account, start));
+       auto itr_end = by_op_idx.lower_bound(boost::make_tuple(account, stop));
+       itr_end--;
```

Comment:
This is unsafe.

Perhaps it's better to make use of `reverse_iterator` and `rend()` to implement this feature.

By the way, please always use `--itr_end` instead of `itr_end--` for better performance.